### PR TITLE
Request URL, referer URL and client IP address for error and exception logs

### PR DESCRIPTION
**Pull request by mirko-pagliai**
Added request URL and referer URL for error logs and the client IP address for error and exception logs

### CODE REVIEW

**Comment by dereuromark** [comment]
Wasn't that already done in https://github.com/cakephp/cakephp/pull/8089 ?

**Comment by mirko-pagliai** [comment]
Sorry, I had not seen.

That PR adds only the referer url to exception logs.

My PR adds:
1) referer url to exception logs;
2) request url and referer url to error logs;
3) client IP to error logs;
4) client IP to exception log.

**Comment by markstory** [comment]
This code needs tests before it can be merged in. There [are also some PHPCS errors](https://travis-ci.org/cakephp/cakephp/jobs/115039576) that need to be corrected.

File: src/Error/BaseErrorHandler.php [hunk 1]
```diff
@@ -270,6 +270,22 @@ protected function _logError($level, $data)
...
+				$referer = $request->env('HTTP_REFERER');
+				if ($referer) {
+					$message .= "\nReferer URL: " . $referer;
+				}
+			
+				$clientIp = $request->clientIp();
+				if	($clientIp && $clientIp !== '::1') {
+					$message .= "\nClient IP: " . $clientIp;
+				}
+			}
```
Comment (markstory) [inline src/Error/BaseErrorHandler.php]: Perhaps we should have a helper method that adds request context to a log line instead of duplicating the code.

**Comment by mirko-pagliai** [comment]
Done. Sorry guys, this is my first PR.

**Comment by markstory** [comment]
Thank you @mirko-pagliai I added some test assertions and removed the duplicate request handling code.

**Comment by mirko-pagliai** [comment]
Thanks to you, you have been very kind.
For the next time, is there a way to see if the code contains errors, before sending the PR (so you do not see my horrors! :P)?

**Comment by ADmad** [comment]
@mirko-pagliai You can locally run the core test cases for classes you have changed. The [manual](http://book.cakephp.org/3.0/en/development/testing.html) has info on how to install and use phpunit.

To check and fix coding standards errors use http://github.com/cakephp/cakephp-codesniffer

**Comment by mirko-pagliai** [comment]
perhaps a downgrading, but this feature does not work anymore. I need to investigate

**Comment by dereuromark** [comment]
Please also check https://github.com/cakephp/cakephp/commit/2e12b89f64b9e35cd1b3f2f4198aa4a095146642#diff-d56691f6e02b95e03d1985514305c47cR226 - there are tests to prevent this regression.

**Comment by mirko-pagliai** [comment]
I'm going back on this thread because I still have problems.
The tests are ok, but in my logs I do not see the client's IP address from a short time (I'm not on localhost).

Only some stupid question to understand: is it right that in case of error and exception methods `handleError()` and `handleException()` are called, and consequently `_logError()` and `_logException()`?

To try something, I'm throwing an exception (eg. `MissingControllerException`). I added an `exit()` to each of these methods, but nothing happens, the exception is normally displayed in the browser and the log is written regularly (no IP address).

Probably I'm wrong in something. I ask just to understand, because I certainly have not understood something.

**Comment by markstory** [comment]
> Only some stupid question to understand: is it right that in case of error and exception methods handleError() and handleException() are called, and consequently _logError() and _logException()?

That is how it is supposed to work. Do you have a custom error handler/exception handler that might be confusing the logic?

**Comment by mirko-pagliai** [comment]
No way, I do not understand.

My `app.php`:
```
    'Error' => [
        'errorLevel' => E_ALL & ~E_DEPRECATED & ~E_USER_DEPRECATED,
        'exceptionRenderer' => 'Cake\Error\ExceptionRenderer',
        'skipLog' => [
            'Cake\Datasource\Exception\RecordNotFoundException',
            'Cake\Routing\Exception\MissingControllerException'
        ],
        'log' => true,
        'trace' => true,
    ],
```
My `bootstrap.php`:
```
$isCli = PHP_SAPI === 'cli';
if ($isCli) {
    (new ConsoleErrorHandler(Configure::read('Error')))->register();
} else {
    (new ErrorHandler(Configure::read('Error')))->register();
}
```
Is there anything else I should check?

**Comment by markstory** [comment]
That looks like it should work and is using the stock error handlers.

**Pull request merged**